website/blog/2021-06-21-use-Java-to-write-Apache-APISIX-plugins.md

```diff
@@ -0,0 +1,233 @@
+import org.apache.apisix.plugin.runner.HttpResponse;
+import org.springframework.stereotype.Component;
+import reactor.core.publisher.Mono;
+
+
+@Component
+public class TokenValidator implements PluginFilter {
+
+    @Override
+    public String name() {
+        return "TokenValidator";
+    }
+
+    @Override
+    public Mono<Void> filter(HttpRequest request, HttpResponse response, PluginFilterChain chain) {
+        return chain.filter(request, response);
+    }
+}
+```
+
+需要继承 `PluginFilter`接口，重写 `name` 和 `filter`函数。
+
+重写 `name` 的返回值，和前面配置 APISIX 的路由属性中 "name" 保持一致。
+
+完整代码如下：
+
+```Java
+package org.apache.apisix.plugin.runner.filter;
+
+import com.google.gson.Gson;
+import org.apache.apisix.plugin.runner.HttpRequest;
+import org.apache.apisix.plugin.runner.HttpResponse;
+import org.springframework.stereotype.Component;
+import reactor.core.publisher.Mono;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Component
+public class TokenValidator implements PluginFilter {
+
+    @Override
+    public String name() {
+        return "TokenValidator";
+    }
+
+    @Override
+    public Mono<Void> filter(HttpRequest request, HttpResponse response, PluginFilterChain chain) {
+        // parse `conf` to json
+        String configStr = request.getConfig(this);
+        Gson gson = new Gson();
+        Map<String, Object> conf = new HashMap<>();
+        conf = gson.fromJson(configStr, conf.getClass());
+
+        // get configuration parameters
+        String token = request.getHeader((String) conf.get("validate_header"));
+        String validate_url = (String) conf.get("validate_url");
+        boolean flag = validate(token, validate_url);
+
+        // token verification results
+        if (!flag) {
+            String rejected_code = (String) conf.get("rejected_code");
+            response.setStatusCode(Integer.parseInt(rejected_code));
+            return chain.filter(request, response);
+        }
+
+        return chain.filter(request, response);
+    }
+
+    private Boolean validate(String token, String validate_url) {
+        //TODO: improve the validation process
+        return true;
+    }
+}
+```
+
+### 4.4 测试
+
+在 Apache APISIX 上配置的上游服务是 httpbin.org，可以访问 Apache APISIX，触发路由，让 Apache APISIX 调用 apisix-java-plugin-runner 去执行 TokenValidator 插件，测试一下 Java 插件效果。
+
+```text
+curl -H 'token: 123456' 127.0.0.1:9080/get 
+{
+ "args": {}, 
+ "headers": {
+ "Accept": "/", 
+ "Host": "127.0.0.1", 
+ "Token": "123456", 
+ "User-Agent": "curl/7.71.1", 
+ "X-Amzn-Trace-Id": "Root=1-60cb0424-02b5bf804cfeab5252392f96", 
+ "X-Forwarded-Host": "127.0.0.1"
+ }, 
+ "origin": "127.0.0.1", 
+ "url": "http://127.0.0.1/get"
+}
+```
+
+## 5. 部署
+
+插件开发完成后，部署操作可以参考 [部署 apisix-java-plugin-runner](https://github.com/apache/apisix-java-plugin-runner/blob/main/docs/how-it-works.md#run)。
+
+## 6. 视频教程
+<iframe 
+    height="350" 
+    width="600" 
+    src="../static/video/2021-06-21-use-Java-to-write-Apache-APISIX-plugins.mp4" 
```

Comment:
Please upload this video to somewhere else, not to this repo.

Reply:
Chiming in. How about uploading this video to APISIX youtube and then making a plug of that here, what do you think ?

Reply:
Not a good idea IMO, because people in China couldn't watch this video, let me help to upload it to CDN.

Reply:
Use this one instead:

https://api7-website-1301662268.file.myqcloud.com/2021-06-21-use-Java-to-write-Apache-APISIX-plugins.mp4

Reply:
thanks